Add api specification support proposal (#422)
kyma-incubator/compass: docs/architecture/api-specs-support.md

# API Specifications Support

## Overview

> This document describes current and planned api specification support in Kyma components.

Reviewer: ```suggestion
This document describes current and planned API specification support in Kyma components.
```

## Currently supported api specification versions and formats

> Following table shows currently supported api specification versions and formats in different components.

Reviewer: ```suggestion
The following table shows currently supported API specification versions and formats in different components.
```

> Last row describes support for extracting api specifications from archive files.

Reviewer: ```suggestion
The last row describes support for extracting API specifications from archive files.
```

> Last row describes support for extracting api specifications from archive files.

Reviewer: ```suggestion
> **NOTE:** The last row describes support for extracting API specifications from archive files.
```

> | | Console UI & Compass UI (rendering) | Rafter (storing) |  Compass UI (uploading) | Compass Backend (storing) | Runtime Agent (storing)

Reviewer 2: How Rafter (storing) and Runtime Agent (storing) differs?

maybe:
1. Rafter (storing) -> Runtime Rafter  (storing). In both tables.
1. will be nice to see red/gree rows in every column about format/version/validation/conversion. By that we will see which column is a most limited one. Think how to visualize that.

Reviewer: Not sure if we should mention the Rafter here. We use it only as a backing service and in terms of our usecases we user nevert interacts with the Rafter directly.

| --- | --- | --- | --- | --- | --- |
| OpenAPI<br>(YAML, JSON) | all formats, v2 and v3 | no validation / conversion | all versions and formats, validation by property | all versions and formats, no validation | all versions, JSON only, no validation and conversion |
| OData<br>(XML, JSON in v4) | XML only, all versions converted to v4 on client side |  no validation / conversion |  all versions, XML only, validation by parsing XML file and checking property | all versions and formats, no validation | all versions and formats, no validation and conversion |     

> | AsyncAPI<br>YAML, JSON | all formats, v2 only | validation and conversion to JSON and v2 | all versions and formats, validation by property | all versions and formats, no validation | all versions, JSON only, no validation and conversion |

Reviewer: To make it consistent with previous rows
```suggestion
| AsyncAPI<br>(YAML, JSON) | all formats, v2 only | validation and conversion to JSON and v2 | all versions and formats, validation by property | all versions and formats, no validation | all versions, JSON only, no validation and conversion |
```

| Archives | no | yes (TAR and ZIP) | no | no | no |

## Planned support for api specification versions and formats

> Following table shows proposed supported api specification versions and formats in different components.

Reviewer: ```suggestion
The following table shows the proposed supported API specification versions and formats in different components.
```

> Rendering would support only a single version of api specifications.

Reviewer: ```suggestion
Rendering would support only a single version of API specifications.
```

> Rafter would be responsible for validating and converting api specifications for Console UI.

Reviewer: ```suggestion
Rafter would be responsible for validating and converting API specifications for Console UI.
```

> Because Compass UI doesn't use Rafter conversion would have to be done in Compass Backend.

Reviewer: ```suggestion
Because Compass UI does not use Rafter conversion would have to be done in Compass Backend.
```

> Preferably Rafter's converters and validators should be exposed as services or libraries that could be used by Compass Backend and Compass UI.

Reviewer 2: 1. Prio one: as a libraries if possible, cause we save specs in our db, and I want to avoid for now additional components in compass cluster.

Author: What can be problematic about libraries is that UI won't be able to use them (unless there would be implementations for different languages).

> Last row describes support for extracting api specifications from archive files. 

Reviewer: ```suggestion
> **NOTE:**  The last row describes support for extracting API specifications from archive files. 
```

| | Console UI & Compass UI (rendering) | Runtime Rafter (storing) |  Compass UI (uploading) | Compass Backend (storing) | Runtime Agent (storing)
| --- | --- | --- | --- | --- |  --- |
| OpenAPI<br>(YAML, JSON) | v3, JSON and YAML | validation and conversion to version supported by renderer | all versions and formats, validation using same logic as rafter | all versions and formats, validation and conversion using same logic as rafter | all versions and formats, no validation and conversion |
| OData<br>(XML, JSON in v4) | v4, JSON |  validation and conversion to version supported by renderer |  all versions and formats, validation using same logic as rafter | all versions and formats, validation and conversion using same logic as rafter | all versions and formats, no validation and conversion |

> | AsyncAPI<br>YAML, JSON | v2, JSON and YAML | validation and conversion to version supported by renderer | all versions and formats, validation using same logic as rafter | all versions and formats, validation and conversion using same logic as rafter | all versions and formats, no validation and conversion |

Reviewer: To make it consistent with previous rows
```suggestion
| AsyncAPI<br>(YAML, JSON) | v2, JSON and YAML | validation and conversion to version supported by renderer | all versions and formats, validation using same logic as rafter | all versions and formats, validation and conversion using same logic as rafter | all versions and formats, no validation and conversion |
```
